### ISSUE TITLE
[🐛] 🔥 Handling multiple projects within the same mobile application

### ISSUE THREAD

**Issue by JCcastagne** [invertase/react-native-firebase#7452]
## Issue

Let's say that I have separate Firebase projects for Client A and Client B. When loading the mobile app, the user is greeted with a screen which they choose if they want to log into Client A or Client B. Afterwards, they'd provide their email/password to sign into that project.

On the web version of our app we are able to switch easily by providing different firebase configs, but I cannot find a way to sign into a different Firebase project on React Native, since it uses the google-services(.json/.plist) that is provided during build time. The issue is that I would like to load (dynamically) the 'google-services' file depending on which company(firebase project) the user selects.

After hours of searching, it seems that I cannot find the exact problem that I am facing. Any help or ideas regarding this topic is greatly appreciated.

** Please note that I am not trying to switch between apps within the same project, what I'm trying to figure out is how to switch between Firebase projects.

---

### Javascript

<details><summary>Click To Expand</summary>
<p>

#### `package.json`:

```json
# N/A
```

#### `firebase.json` for react-native-firebase v6:

```json
# N/A
```

</details>

### iOS

<details><summary>Click To Expand</summary>
<p>

#### `ios/Podfile`:

- [ ] I'm not using Pods
- [x] I'm using Pods and my Podfile looks like:

```ruby
# N/A
```

#### `AppDelegate.m`:

```objc
// N/A
```

</p>
</details>

---

### Android

<details><summary>Click To Expand</summary>
<p>

#### Have you converted to AndroidX?

- [ ] my application is an AndroidX application?
- [ ] I am using `android/gradle.settings` `jetifier=true` for Android compatibility?
- [ ] I am using the NPM package `jetifier` for react-native compatibility?

#### `android/build.gradle`:

```groovy
// N/A
```

#### `android/app/build.gradle`:

```groovy
// N/A
```

#### `android/settings.gradle`:

```groovy
// N/A
```

#### `MainApplication.java`:

```java
// N/A
```

#### `AndroidManifest.xml`:

```xml
<!-- N/A -->
```

</p>
</details>

---

## Environment

<details><summary>Click To Expand</summary>
<p>

**`react-native info` output:**

```
// this command does not exist in Expo-managed projects, instead, here is info I believe would be usefull:
    "@react-native-firebase/app": "^18.3.0",
    "@react-native-firebase/auth": "^18.3.0",
    "@react-native-firebase/database": "^18.3.0",
    "@react-native-firebase/firestore": "^18.3.0",
    "@react-native-firebase/storage": "^18.3.0",
    "expo": "~49.0.7",
    "react-native": "0.72.3",
```

- **Platform that you're experiencing the issue on**:
  - [ ] iOS
  - [ ] Android
  - [ ] **iOS** but have not tested behavior on Android
  - [ ] **Android** but have not tested behavior on iOS
  - [X] Both
- **`react-native-firebase` version you're using that has this issue:**
  - `18.3.0`
- **`Firebase` module(s) you're using that has the issue:**
  - `n/a`
- **Are you using `TypeScript`?**
  - `N`

</p>
</details>

---

**Comment by mikehardy** [comment]
Hi there 👋  there are constraints to using some of the native firebase SDK modules, and those constraints are that some of the modules load *immediately* as the app starts, and must configure themselves *immediately*, which is why we have the google services json/plist as requirements.

For instance, there are many crashlytics, performance, and analytics things that are measured and maybe even sent before javascript ever gets control in order to maybe override a project, right? (some discussion upstream in their docs: https://firebase.google.com/docs/projects/multiprojects)

However, you may initialize many of the firebase modules with different project configs at runtime, what you need to look for is which modules support that. Here's the list from our documentation and...I believe we support multiple projects ("secondary apps") here where ever they are supported upstream / in the native SDKs: https://rnfirebase.io/app/usage#secondary-apps

As long as you stick to those modules, you can initalize a "secondary app" by creating a new instance of the module at runtime with a different firebase config and it should work.

You'll want to have a primary config that is "generic" I suppose, for the things you cannot change, and if you really must change them somehow (for instance, analytics...) you'll need to use a pure javascript solution such as either firebase-js-sdk or some other thing)

That's the full answer I think, and the best we can do

**Comment by JCcastagne** [comment]
Hi Mike, thank you for the quick response.

> However, you may initialize many of the firebase modules with different project configs at runtime (...)

> As long as you stick to those modules, you can initalize a "secondary app" by creating a new instance of the module at runtime with a different firebase config and it should work.

I have tried initializing a secondary app that uses those supported modules listed above with another project's config, but get the following error:
`NativeFirebaseError: [app/unknown] Configuration fails. It may be caused by an invalid GOOGLE_APP_ID in GoogleService-Info.plist or set in the customized options.`

This error seemingly is because Firebase only looks at the initial google-services(.json/.plist) file at the beginning of build/run time while initializing the default app, which of course would throw an error since I'm trying to load another project's config.

Hope this clarifies exactly what the issue is.

**Comment by mikehardy** [comment]
🤔  without more information along the lines of 

...I can't say exactly

You may take inspiration from our e2e test suite though, where this sort of thing is tested all the time and works: https://github.com/invertase/react-native-firebase/blob/2c1127a630dd7e800631693b65f9d0e8a63e798b/packages/app/e2e/app.e2e.js#L70-L81

The config may be ios or android specific, necessitating a platform check prior to loading it in as the config - we do that, using configs downloaded from the firebase console for a secondary test project. It definitely works

**Comment by JCcastagne** [comment]
I don't believe that creating an example would help with my question, but please let me know if you'd still need one if I don't explain it better this time:

The help I'm looking for is to switch between projects, but everything I search for or when I get replies, it's people sending me over to the same link which is "how to switch **between apps within** a project). I'm looking to actually **switch between projects**, as in databases.

In other words, I would like to be able to connect to Firebase using different google-services.json/.plist files, depending which project I want during run time.

**Comment by mikehardy** [comment]
tWith regards to entire projects it is not possible while using the native SDKs that I know of. And you have solid observational evidence that it does not work - I've never tried it, so your experimental result is the definitive result at this point.

However, firebase-js-sdk will run in a react-native context. It does not allow deeper integration with the device (like, certain types of messaging, crashlytics, performance, analytics etc - the things that don't work with secondary apps because they are natively integrated) but it can perhaps work for your business requirement.
